File: java/socketd/src/main/java/org/noear/socketd/transport/core/impl/SessionBase.java
package org.noear.socketd.transport.core.impl;

import org.noear.socketd.transport.core.Channel;
import org.noear.socketd.transport.core.Session;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 会话基类
 *
 * @author noear
 */
public abstract class SessionBase implements Session {
    protected final Channel channel;

    private final String sessionId;

    public SessionBase(Channel channel) {
        this.channel = channel;
        this.sessionId = generateId();
    }

    /**
     * 会话的附件与通道的各自独立
     */
    private Map<String, Object> attrMap;

    @Override
    public Map<String, Object> attrMap() {
        if (attrMap == null) {
            attrMap = new ConcurrentHashMap<>();
        }

        return attrMap;
    }

    /**
     * 是有属性
     *
     * @param name 名字
     */
    @Override
    public boolean attrHas(String name) {
        if (attrMap == null) {
            return false;
        }

        return attrMap.containsKey(name);
    }

    /**
     * 获取附件
     */
    @Override
    public <T> T attr(String name) {
        if (attrMap == null) {
            return null;
        }

        return (T) attrMap.get(name);
    }

    /**
     * 获取属性或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    @Override
    public <T> T attrOrDefault(String name, T def) {
        T tmp = attr(name);
        if (tmp == null) {
            return def;
        } else {
            return tmp;
        }
    }

    /**
     * 设置附件
     */
    @Override
    public <T> Session attrPut(String name, T value) {
        if (attrMap == null) {
            attrMap = new ConcurrentHashMap<>();
        }
        attrMap.put(name, value);
        return this;
    }

    @Override
    public String sessionId() {
        return sessionId;
    }

    @Override
    public long liveTime() {
        return channel.getLiveTime();
    }

    protected String generateId() {
        return channel.getConfig().genId();
    }

    //不要加 hashCode, equals 重写！
}

File: python/socketd/socketd/transport/core/impl/SessionBase.py
from abc import ABC
from typing import Any, Dict
from socketd.transport.core.Session import Session
from socketd.transport.core.Channel import Channel
from socketd.utils.MapUtils import MapUtils


class SessionBase(Session, ABC):
    def __init__(self, channel: Channel):
        self._channel = channel
        self._session_id = self.generate_id()
        self._attrMap:Dict[str, Any] = None

    def attr_map(self) -> Dict[str, Any]:
        if self._attrMap is None:
            self._attrMap = {}

        return self._attrMap

    def attr_has(self, name:str) -> bool:
        if self._attrMap is None:
            return False
        else:
            return self._attrMap.__contains__(name)

    def attr_del(self, name: str):
        if self._attrMap is not None:
            MapUtils.remove(self._attrMap, name)

    def attr(self, name: str) -> Any:
        if self._attrMap is None:
            return None

        return self._attrMap.get(name)

    def attr_or_default(self, name: str, defVal: Any) -> Any:
        tmp = self.attr(name)
        if tmp is None:
            return defVal
        else:
            return tmp

    def attr_put(self, name: str, val: Any) -> None:
        if self._attrMap is None:
            self._attrMap = {}
        self._attrMap[name] = val

    def session_id(self) -> str:
        return self._session_id

    def is_active(self) -> bool:
        return self.is_valid() and self.is_closing() == False

    def live_time(self)->int:
        return self._channel.get_live_time()

    def generate_id(self) -> str:
        return self._channel.get_config().gen_id()

